pkg/querier/frontend/frontend_querier_queues.go

```diff
@@ -0,0 +1,216 @@
+	queriers    map[string]struct{}
+	maxQueriers int
+
+	// Seed for shuffle sharding of queriers. This seed is based on userID only and is therefore consistent
+	// between different frontends.
+	seed int64
+
+	// Points back to 'users' field in queues. Enables quick cleanup.
+	index int
+}
+
+func newUserQueues(maxUserQueueSize int) *queues {
+	return &queues{
+		userQueues:         map[string]*userQueue{},
+		users:              nil,
+		maxUserQueueSize:   maxUserQueueSize,
+		querierConnections: map[string]int{},
+		sortedQueriers:     nil,
+	}
+}
+
+func (q *queues) len() int {
+	return len(q.userQueues)
+}
+
+func (q *queues) deleteQueue(userID string) {
+	uq := q.userQueues[userID]
+	if uq == nil {
+		return
+	}
+
+	delete(q.userQueues, userID)
+	q.users[uq.index] = ""
+
+	// Shrink users list size if possible. This is safe, and no users will be skipped during iteration.
+	for ix := len(q.users) - 1; ix >= 0 && q.users[ix] == ""; ix-- {
+		q.users = q.users[:ix]
+	}
+}
+
+// Returns existing or new queue for user.
+// MaxQueriers is used to compute which queriers should handle requests for this user.
+// If maxQueriers is <= 0, all queriers can handle this user's requests.
+// If maxQueriers has changed since the last call, queriers for this are recomputed.
+func (q *queues) getOrAddQueue(userID string, maxQueriers int) chan *request {
+	// Empty user is not allowed, as that would break our users list ("" is used for free spot).
+	if userID == "" {
+		return nil
+	}
+
+	if maxQueriers < 0 {
+		maxQueriers = 0
+	}
+
+	uq := q.userQueues[userID]
+
+	if uq == nil {
+		uq = &userQueue{
+			ch:    make(chan *request, q.maxUserQueueSize),
+			seed:  getSeedForUser(userID),
+			index: -1,
+		}
+		q.userQueues[userID] = uq
+
+		// Add user to the list of users... find first free spot, and put it there.
+		for ix, u := range q.users {
+			if u == "" {
+				uq.index = ix
+				q.users[ix] = userID
+				break
+			}
+		}
+
+		// ... or add to the end.
+		if uq.index < 0 {
+			uq.index = len(q.users)
+			q.users = append(q.users, userID)
+		}
+	}
+
+	if uq.maxQueriers != maxQueriers {
+		uq.maxQueriers = maxQueriers
+		uq.queriers = q.selectQueriersForUser(uq.seed, q.sortedQueriers, maxQueriers)
+	}
+
+	return uq.ch
+}
+
+// Finds next queue for the querier. To support fair scheduling between users, client is expected
+// to pass last user index returned by this function as argument. Is there was no previous
+// last user index, use -1.
+func (q *queues) getNextQueueForQuerier(lastUserIndex int, querier string) (chan *request, string, int) {
+	uid := lastUserIndex
+
+	for iters := 0; iters < len(q.users); iters++ {
+		uid = uid + 1
+
+		// Don't use "mod len(q.users)", as that could skip users at the beginning of the list
+		// for example when q.users has shrunk since last call.
+		if uid >= len(q.users) {
+			uid = 0
+		}
+
+		u := q.users[uid]
+		if u == "" {
+			continue
+		}
+
+		q := q.userQueues[u]
+
+		if q.queriers != nil {
```

Comment:
i'm not sure how to quantify the effect, but this could unfairly skip users who don't happen to be assigned to the querier currently asking for a query.

Reply:
I don’t quite follow here... skipping users who use different queriers is the goal of this PR.